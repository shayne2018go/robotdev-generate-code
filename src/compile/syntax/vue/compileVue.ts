import { ICodeSchema } from '@/types/ICodeSchema';
import compilePages from './compilePages';
import compileRouter, { VueRoute } from './compileRouter';
import compileComponents, { VueComponent } from './compileComponents';
import compileFunctions, { VueFunction } from './compileFunctions';

export interface VueCompileOptions {
  routes: VueRoute[]; // 路由相关
  components: VueComponent[]; // 组件相关
  functions: VueFunction[]; // 函数相关
}

function compileVue(codeSchema: ICodeSchema) {
  const vueCompileOptions = {
    routes: [],
    functions: [],
    components: [],
  } as VueCompileOptions;

  const { tokens: routerTokens, routes } = compileRouter(codeSchema);
  vueCompileOptions['routes'] = routes;

  const { tokens: functionTokens, functions } = compileFunctions(codeSchema);
  vueCompileOptions['functions'] = functions;

  const { tokens: componentTokens, components } = compileComponents(codeSchema, vueCompileOptions);
  vueCompileOptions['components'] = components;

  const { tokens: pageTokens } = compilePages(codeSchema);

  const tokens = routerTokens.concat(componentTokens).concat(functionTokens).concat(pageTokens);

  return { tokens };
}

export default compileVue;
