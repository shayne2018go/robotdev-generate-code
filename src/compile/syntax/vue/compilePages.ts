import createToken from '@/compile/config/createToken';
import { Compile } from '@/types/compile/token';
import { VueGlobalCtx } from './compileVue';
import { BUILT_IN_IMPORTED_PACKAGE, PAGE_DIR } from './const/config';
import { compileScript, compileStyle, compileTemplate } from './sfc';
import { componentEmitsDataStore, componentSlotsDataStore } from './shared/store';
import { nodesDataStore } from './shared/store/nodes';
import { propertiesDataStore } from './shared/store/properties';

export type CompileCurrentCtx = {
  global: VueGlobalCtx;
  scope: {
    current: ParsingCurrentResult;
  };
};

export interface ParsingCurrentResult {
  data: CodeSchema.Component | CodeSchema.Page;
  nodesStore: ReturnType<typeof nodesDataStore>;
  variablesStore: ReturnType<typeof propertiesDataStore>;
  lifeCyclesStore: ReturnType<typeof componentEmitsDataStore>;
  slotsStore: ReturnType<typeof componentSlotsDataStore>;
  emitsStore: ReturnType<typeof componentEmitsDataStore>;
  propsStore: ReturnType<typeof propertiesDataStore>;
  importComponents: GlobalContext.Component[];
  importFunctions: GlobalContext.Function[];
}

// interface ParsingPageResult {
//   page: CodeSchema.Page;
//   nodesStore: ReturnType<typeof nodesDataStore>;
//   variablesStore: ReturnType<typeof propertiesDataStore>;
//   importComponents: GlobalContext.Component[];
//   importFunctions: GlobalContext.Function[];
// }

// var

function compilePages(codeSchema: CodeSchema.Project, vueGlobalCtx: VueGlobalCtx): { tokens: Compile.Token[] } {
  const tokens = [] as Compile.Token[];

  const { directories = [] } = codeSchema;

  codeSchema.pages.forEach((page) => {
    // 1 编译页面代码字符串（token）
    const { token } = compilePage(page, vueGlobalCtx);
    // 2 编译页面地址 ps： 根据directories
    const filePath = getPageFilePath(page, vueGlobalCtx);

    tokens.push(createToken(filePath, token));
  });

  return { tokens };
}

function compilePage(page: CodeSchema.Page, ctx: VueGlobalCtx) {
  const parsingComponentResult = parsingCurrent(page, ctx);

  const currentPageCompileOptions: CompileCurrentCtx = {
    global: ctx,
    scope: {
      current: parsingComponentResult,
    },
  };

  const { token: templateToken } = compileTemplate(page, currentPageCompileOptions);
  const { token: scriptToken } = compileScript(page, currentPageCompileOptions);
  const { token: styleToken } = compileStyle(page, currentPageCompileOptions);

  const token = templateToken + scriptToken + styleToken;

  return { token };
}

export function parsingCurrent<T extends CodeSchema.Component | CodeSchema.Page>(
  data: T,
  ctx: VueGlobalCtx
): ParsingCurrentResult {
  const importComponents: GlobalContext.Component[] = parsingCurrentImportedComponents(data, ctx);

  const buitinImportedFunctions: GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>[] =
    ctx.functionsStore.findAll().filter((func) => {
      return func.source?.packageName && BUILT_IN_IMPORTED_PACKAGE.includes(func.source?.packageName);
    }) || [];

  const importFunctions: GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>[] = [
    {
      id: 'open', // 函数id
      key: 'open', // 函数名称
      source: {
        filePath: 'src/system/action.ts', // 本地路径
        exportName: 'open', // 导出名 (不能为空 默认default )
      },
    } as GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>,
    {
      id: 'queryImage', // 函数id
      key: 'queryImage', // 函数名称
      source: {
        filePath: 'src/system/action.ts', // 本地路径
        exportName: 'queryImage', // 导出名 (不能为空 默认default )
      },
    } as GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>,
    {
      id: 'assignment', // 函数id
      key: 'assignment', // 函数名称
      source: {
        filePath: 'src/system/action.ts', // 本地路径
        exportName: 'assignment', // 导出名 (不能为空 默认default )
      },
    } as GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>,
  ].concat(buitinImportedFunctions);

  const nodesStore = nodesDataStore(data.nodes, ctx);
  const emitsStore = componentEmitsDataStore(isComponent(data) ? data.emits : []);
  const slotsStore = componentSlotsDataStore(isComponent(data) ? data.slots : []);
  const propsStore = propertiesDataStore(isComponent(data) ? data.props : []);
  const variablesStore = propertiesDataStore(data.variables || []);
  const lifeCyclesStore = ctx.eventsStore;

  const result = {
    data,
    nodesStore,
    variablesStore,
    lifeCyclesStore,
    importComponents,
    importFunctions,
    emitsStore,
    slotsStore,
    propsStore,
  } as ParsingCurrentResult;

  return result;
}

function isPage(data: Record<string, any>): data is CodeSchema.Page {
  return data.meta || data.route;
}

function isComponent(data: Record<string, any>): data is CodeSchema.Component {
  return !data.meta && !data.route;
}

function getPageFilePath(page: CodeSchema.Page, vueGlobalCtx: VueGlobalCtx): string {
  // const dirPath: CodeSchema.Directory[] = [];

  // let curId: string | null = page.id;

  // while (curId) {
  //   const current = directories.find((d) => d.id === curId);
  //   if (current) {
  //     dirPath.unshift(current);
  //     curId = current.parentId;
  //   } else {
  //     break;
  //   }
  // }

  // // vue-router 的path
  // const path = `/${dirPath.map((p) => p.key).join('/')}`;

  // // 文件的绝对路径
  // const filePath = `${PAGE_DIR}${path}.vue`;
  const page_protocol = vueGlobalCtx.pagesStore.getPage(page.id);
  if (!page_protocol) {
    throw new Error(`Page ${page.id} is not found`);
  }
  const filePath = page_protocol.source?.filePath || `${PAGE_DIR}/${page_protocol.key}.vue`;

  return filePath;
}

function parsingCurrentImportedComponents<T extends CodeSchema.Component | CodeSchema.Page>(
  data: T,
  ctx: VueGlobalCtx
): GlobalContext.Component[] {
  const importedComponents = [] as GlobalContext.Component[];

  data.nodes.forEach((node) => {
    const isExist = ctx.componentsStore.getCmpt(node.tagId);

    if (isExist && isExist.source?.filePath) {
      importedComponents.push(isExist);
    }
  });
  return importedComponents;
}

export default compilePages;
