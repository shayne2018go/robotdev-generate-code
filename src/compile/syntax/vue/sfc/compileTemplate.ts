import { Compile } from '@/types/compile/token';
import { INode } from '@/types/view';
import { CompilePageCtx, CompilePageOptions } from '../compilePages';
import { ICS_Page } from '@/types/page';

const WHITE_TAG = ['each', 'cond', 'slot', 'tpl'];

// interface CompileTemplateOptions {}

function compileTemplate(page: ICS_Page, compileCtx: CompilePageCtx): { token: string } {
  // listToTree
  // tag props events children
  const token = '';

  return { token };
}

function getTag(tagId: string): string {
  return '';
}

export default compileTemplate;
