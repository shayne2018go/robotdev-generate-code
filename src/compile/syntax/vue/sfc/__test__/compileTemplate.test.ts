import { describe, expect, it } from 'vitest';
import compileTemplate from '../compileTemplate';
import { parsingVueCompileOptions, buildGlobalCtx } from '../../compileVue';
import codeSchema from '@/__test__/__fixture__/CodeSchema';
import { CompilePageCtx, parsingPage } from '../../compilePages';
import templateToken from './__fixture__/template-tokens.json';

describe('compileTemplate', () => {
  // 解析相关依赖协议
  const vueCompileOptions = parsingVueCompileOptions(codeSchema);

  // 构建全局上下文
  const vueGlobalCtx = buildGlobalCtx(vueCompileOptions);

  const page = codeSchema.pages[3];

  const parsingPageResult = parsingPage(page, vueGlobalCtx);

  const currentPageCompileOptions: CompilePageCtx = {
    global: vueGlobalCtx,
    scope: {
      page: parsingPageResult,
    },
  };

  it('tempalte', () => {
    const { token } = compileTemplate(page, currentPageCompileOptions);
    debugger;

    expect(token).toMatchSnapshot();
    expect(token).toEqual(templateToken);
  });
});
