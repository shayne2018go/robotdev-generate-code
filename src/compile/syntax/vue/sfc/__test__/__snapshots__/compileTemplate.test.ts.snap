// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compileTemplate > tempalte 变量 事件 1`] = `
"<template><a-card><a-form><a-form-item :label=\\"nodesState.aFormItem.label\\"><a-input :value=\\"nodesState.aInput.value\\" @change=\\"(...args) => {
  nodesState.aInput.change(...args);
}\\"></a-input></a-form-item><a-form-item :label=\\"nodesState.aFormItem_2.label\\"><a-input-number @change=\\"(...args) => {
  nodesState.aInputNumber.change(...args);
}\\"></a-input-number></a-form-item><a-form-item :label=\\"nodesState.aFormItem_3.label\\" :labelAlign=\\"nodesState.aFormItem_3.labelAlign\\"><a-button :type=\\"nodesState.aButton.type\\" :size=\\"nodesState.aButton.size\\" :shape=\\"nodesState.aButton.shape\\" @click=\\"(...args) => {
  nodesState.aButton.click(...args);
}\\">{{nodesState.text.text}}</a-button><a-button :shape=\\"nodesState.aButton_2.shape({})\\" @click=\\"(...args) => {
  nodesState.aButton_2.click(...args);
}\\">{{nodesState.text_2.text}}</a-button></a-form-item></a-form></a-card></template>"
`;

exports[`compileTemplate > tempalte 循环 判断 1`] = `
"<template><template v-for=\\"(each_item, each_index) in nodesState.each.data\\"><template v-for=\\"(each_2_item, each_2_index) in nodesState.each_2.data\\"><template v-for=\\"(each_3_item, each_3_index) in nodesState.each_3.data\\"><template v-if=\\"nodesState.cond.value({
  each_item: each_item,
  each_2_item: each_2_item,
  each_3_item: each_3_item
})\\"><a-card :title=\\"each_3_item?.ID\\">{{each_3_item?.yonghuming}}{{nodesState.text_2.text({
  each_item: each_item,
  each_2_item: each_2_item,
  each_3_item: each_3_item
})}}{{each_3_item?.yonghunianling}}</a-card>{{nodesState.text_4.text}}</template></template></template><template v-for=\\"(each_4_item, each_4_index) in each_item\\"><input :type=\\"nodesState.input.type\\" :placeholder=\\"each_4_item\\" /></template></template></template>"
`;
